Add content security policy support

xmorave2 opened the pull request:
This is first "somehow" working implementation. There is still some work on it, but I am sharing it for getting some feedback .

TODO:
* [x] Implement report only mode
* [x] Use headscript view helper or nonce helper for all inline scripts
* [x] Remove usage of HTML event attributes
    * [x] oninvalid
    * [x] onclick
* [x] Generate header in own class for better extensibility/overide of CSP settings
* [x] Laminas migration
* [x] Update to reflect merge of https://github.com/laminas/laminas-http/pull/33 
* [x] Create own CSP report only header class as temporary solution while the laminas-http is not merged and released yet
* [x] Fix Whoops handler
* [x] Use report only mode in development mode (when report only is available)
* [x] Finish updating [wiki page](https://vufind.org/wiki/administration:security:content_security_policy) after merging PR.

demiankatz commented on module/VuFindTheme/src/VuFindTheme/View/Helper/HeadScript.php:
```diff
@@ -103,6 +113,7 @@ public function itemToString($item, $indent, $escapeStart, $escapeEnd)
             }
         }
 
+        $item->attributes['nonce'] = $this->cspNonce;
```
I believe that you could also use `$this->getView()->plugin('cspnonce')` here if you want to avoid the extra injection work... though if you prefer to be more explicit, that's fine with me. Just wanted to point out the more concise alternative in case you found it more appealing.

xmorave2 replied:
I like the explicit injection here... there is possibility that we can avoid using cspnonce view helper even in templates (= inject all sources through other helpers) and than we can even remove nonce helper.

demiankatz replied:
Sounds good to me!

demiankatz commented on module/VuFind/src/VuFind/Controller/AbstractBase.php:
```diff
@@ -164,6 +174,27 @@ protected function createViewModel($params = null)
+        $headers = $response->getHeaders();
+        $headers->addHeaderLine(
+            'Content-Security-Policy',
+            "script-src 'strict-dynamic' 'nonce-$this->nonce' 'unsafe-inline' "
```
Do we eventually want to make some or all of this configurable?

demiankatz replied:
If it's too complex to put in an .ini file due to dynamic elements like the nonce, maybe there should be a service that builds the string so it's easier to override.

xmorave2 replied:
I agree that some service which could generate the header would be fine, maybe we can even find some library, which could make it for us... like https://github.com/paragonie/csp-builder but maybe this library is way too complex for us...

xmorave2 replied:
I added the todo item for this. Some kind of flexibility does make sense here...

demiankatz replied:
The csp-builder library certainly looks like a possibility -- I'd rather use a well-established library than build something new, as long as the library meets our needs and doesn't overcomplicate things too much. In this case, it looks like we have the option of using a config file to initialize the CSP, or we could just build a factory to do the work programmatically.

xmorave2 replied:
I agree, I looked into other library too: https://github.com/BePsvPT/secure-headers, but it looks even more complex then paragonie/csp-builder, so I think we can give a try to it.
 
Just one question remains here: If it is better to do in configuration, or programatically...

If we go with configuration, it will probably have more flexibility for non-programer implementers, but it is in JSON, and I think we can't use configuration inheritence then... So there will be a bit of inconsistence with other configuration.

demiankatz replied:
I think if we load the configuration through the factory, then we can address both scenarios: by default, we can load a configuration, but a user could override the factory to do a programmatic configuration instead if they wanted to. You're right that JSON limits our ability to allow configurations to inherit from one another, but we can at least allow file overriding through the local directory by selecting the file using `\VuFind\Config\Locator::getConfigPath()`.

demiankatz commented on themes/bootstrap3/templates/breadcrumbs/multi.phtml:
```diff
@@ -17,4 +17,4 @@
   <li class="active" title="<?=$this->layout()->title ?>"><?=$this->truncate($this->layout()->title, 100) ?></li>
 <?php endif; ?>
-<script>$('.dropdown-toggle').dropdown()</script>
+<?php $this->headScript()->appendScript("$('.dropdown-toggle').dropdown()");
```
I suspect that this script is inline for a reason, so it's probably better to use the inline script helper rather than the head script helper. I'm also not sure if this template is actually used, so it may not really matter. Either way, perhaps we should take care of eliminating explicit script tags in a separate pull request, since that's a worthwhile effort independent of CSP implementation, and we can probably get it merged sooner.

xmorave2 replied:
changed in follow-up

demiankatz commented on themes/bootstrap3/templates/record/storageretrievalrequest.phtml:
```diff
@@ -93,7 +93,7 @@
 </form>
 
-<script type="text/javascript">
+<script type="text/javascript" nonce="<?=$this->cspnonce()?>">
```
This is another area where it may be better to use the inlinescript helper rather than modify the explicitly coded script tag.

xmorave2 replied:
changed in follow-up

demiankatz commented:
Thanks, @xmorave2; I'm not deeply familiar with CSPs yet, but hopefully this will serve as a good opportunity to learn more. In any case, I've left a few initial comments for you; please let me know when you'd like me to do some hands-on testing.

xmorave2 commented:
@demiankatz Thanks for looking at my code Demian, I added some more commits, but this is still WIP of course..

Note I had to update jquery library, as version 2.1.1 included in current code is using eval in ajax calls - this is problem when you want to have a secure CSP settings.... This may need some more investigation to ensure it is not breaking anything.
Here is upgrade documentation for reference: https://jquery.com/upgrade-guide/3.0/

demiankatz commented:
Thanks for the progress, @xmorave2! Would you mind opening a separate PR with the jQuery upgrade? That will make it easier to test that change in isolation, and to discuss/resolve any problems that arise without making this PR harder to read. Once open, perhaps @crhallberg can weigh in with his thoughts over there.

demiankatz commented on themes/bootstrap3/templates/search/reservessearch.phtml:
```diff
@@ -18,7 +18,7 @@
     <input class="btn btn-primary" type="submit" name="submit" value="<?=$this->transEsc("Find")?>"/>
   </form>
-  <script type="text/javascript">$("#reservesSearchForm_lookfor").focus()</script>
+  <?php $this->headScript()->appendScript("$('#reservesSearchForm_lookfor').focus()");?>
```
Looks like this is another inline script (unintentionally?) changed to a head script in the refactoring. I think I may have just missed this on my first review.

xmorave2 replied:
Fixed in follow-up

demiankatz replied:
Thanks!

xmorave2 commented:
So there are some more small enhancement, and a new pull request for jQuery update.

demiankatz commented:
@xmorave2, have you looked at the CSP support built into Zend\Http? I'm working on updating dependencies (see #1531) and noticed that the Zend\Http changelog (https://github.com/zendframework/zend-http/blob/master/CHANGELOG.md) references recent improvements in that area... perhaps that is of some use. I expect to merge #1531 in a day or two as long as I don't run into problems during testing.

xmorave2 commented:
Good idea @demiankatz, I used zend http headers for small and primitive header generator, and it works OK, I have even implement basic setting through ini file. The only thing I am missing is ContentSecurityPolicyReportOnly implemantation in zend http module... But should be easy to implement, so I'll try to add a pull request to laminas-http.

demiankatz commented:
Thanks, @xmorave2, sounds good! I'm hoping to merge #1537 fairly early in the development of VuFind 7.0, so if we need to upgrade to a newer laminas-http to merge this, it shouldn't be a major problem. If you do open a pull request there, feel free to link it here so we can watch its status.

xmorave2 commented:
The pull request for Laminas is here: https://github.com/laminas/laminas-http/pull/33

demiankatz commented:
@xmorave2, let me know if you'd like any help getting this branch up to date with Laminas while we wait for news on the status of your pull request. (I also notice that your Laminas PR is failing the Travis build; you might want to review that to see if fixing the underlying issue will expedite the merge process... again, let me know if I can help!)

xmorave2 commented:
@demiankatz I migrated the code to Laminas. Regarding the failed build on my pull request for laminas-http, the error reported by travis is the same as on master branch. So it should not be related with my pull request hopefully.

demiankatz commented:
Thanks, @xmorave2, sounds good. Do you need me to do anything else here at this time, or are you waiting on the PR before proceeding?

xmorave2 commented:
I continued on this work. I believe we are close.
There are only two things to fix:
1. whoops handler
2. report only mode - if the pull request would not be accepted into laminas-http it could be easily implemented in vufind itself - in fact it just inherit from csp header and only change the header name

demiankatz commented:
Thanks, @xmorave2! Can you tell me more about the Whoops issue? Is it something I might be able to help with? (Sorry if this is something I should remember -- but from the conversation history on this PR, I can't figure out exactly what that checkbox means).

I'll also leave a comment on the laminas pull request (and a checkbox here to remind me to watch it further).

xmorave2 commented:
@demiankatz: Sorry for not being clear here. I added the todo item quickly when I realized the problem to not forget it.

There is an explanation: When I use CSP, the whoops handler page does not look pretty: browser does not use any css or js. It is because there is no option to pass the nonce to whoops itself. Its template looks like this: https://github.com/filp/whoops/blob/master/src/Whoops/Resources/views/layout.html.php#L14

But we probably could somehow say to whoops to use other template. Or maybe we could just not use CSP when whoops error handler is activated. I am not yet sure how to fix it....

demiankatz commented:
Thanks, @xmorave2, that makes sense. Since Whoops is only enabled in development mode, the actual security concerns here are not as broad as in other parts of VuFind, so if we have to disable CSP in Whoops to make things work, I don't see that as being a major problem... but of course if we can make things work more consistently across the board, that would be preferable.

Let me know if you'd like help digging into the Whoops module in more detail; I haven't given it a particularly close look myself, but I can set aside some time if this requires deeper discussion.

xmorave2 commented:
@demiankatz I done it that way. The header generator class now detect if VuFind is running iin development mode and then use less strict setting.

demiankatz commented:
Thanks, @xmorave2, that makes sense; once the "report only" header is available, would it make sense to use that in development mode, so we could apply consistent rules but not have blocking of content?

xmorave2 commented:
> Thanks, @xmorave2, that makes sense; once the "report only" header is available, would it make sense to use that in development mode, so we could apply consistent rules but not have blocking of content?

@demiankatz: It make sense to me, great idea, I'll refactor it that way

xmorave2 commented:
As the release of 7.0 is closer and that I am not feeling very optimistic about my laminas-http PR to be merged or even released in time. I decided to create CSP report only header class in VuFind's codebase. We could easily use the one from laminas when it is released just by change of one 'use' line (and deleting on class file).

I also refactored CspHeaderGenerator, to simply use report only mode on development as you suggested @demiankatz.

I also made a little polishing of code/fix typos and so on.

demiankatz commented on themes/bootstrap3/templates/install/fixsecurity.phtml:
```diff
@@ -18,7 +18,7 @@
   </p>
   <p><i>Do you still wish to proceed with enabling enhanced security in the database?</i></p>
-  <form method="post" action="<?=$this->url('install-fixsecurity')?>">
+  <form method="post" action="<?=$this->url('install-fixsecurity')?>">ag
```
Is this a typo?

xmorave2 replied:
yes, it is, fixed

demiankatz commented on module/VuFind/src/VuFind/Controller/AbstractBase.php:
```diff
@@ -179,6 +180,35 @@ protected function createViewModel($params = null)
+     *
+     * @throws Exception\DomainException
+     */
+    public function onDispatch(MvcEvent $e)
```
Not every controller extends AbstractBase in VuFind (currently, most do, but that may change with future refactoring when time is found to move away from serviceLocator-dependent helper methods in favor of more disciplined dependency injection). I wonder if it would be more future-proof to set up this functionality through the VuFind\Bootstrapper rather than relying on specific inheritance patterns.

(This is just a suggestion; it will probably work as-is for now... but I just want to be sure we're applying this as thoroughly and completely as possible).

xmorave2 replied:
No problem, I refactored into Bootstrapper. I think, it should be ideally implemented as PSR15 middleware. But it would need much more infrastructure and refactoring definitely out of the scope of this PR. But may be worth thinking about it for the future.

demiankatz replied:
Yes, that might be a good project for VuFind 8.0; feel free to open a JIRA "architecture" ticket if you want a place to discuss the idea further.

demiankatz commented on config/vufind/contentsecuritypolicy.ini:
```diff
@@ -0,0 +1,30 @@
+base-uri[] = "'self'"
+; Set URI which should browser use to report CSP violation, you should provide url
+; when you enable report_only
+report-uri[] = 'https://example.report-uri.com'
```
Should these be commented out by default so we don't try to send things to invalid URLs by default? (I'm not sure I fully understand their purpose, so please correct me if it makes no difference).

xmorave2 replied:
You are right. When there is report-uri directive, the browser tries to send JSON-formatted report of any content security policy violence to defined uri. So i commented it out.

It could be possible to create a handler for such reporting in VuFind itself, but it is out of the scope of this PR. And may not be worth to spend time on it, as there are some at least two main possibilities how to get these report - commercial service report-uri.com and seek-oss csp-server as opensource alternative, written in Node.js, with Elasticsearch as backend - so relatively heavy application...

demiankatz replied:
I agree that building this functionality into VuFind is probably overkill, at least for the moment. Of course we can always consider it if there's a strong need in the future, but we might as well start by getting the basic functionality in place.

Perhaps it's worthwhile to mention these reporting tool options in the comments in the .ini file to help people understand how to use the setting.

demiankatz replied:
I've added notes about the available software/service in the wiki.

demiankatz commented on config/vufind/contentsecuritypolicy.ini:
```diff
@@ -0,0 +1,30 @@
+; Settings for Content Security Policy header, you can learn more here:
+; https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
+[CSP]
+; It is possible to use CSP only in report mode. See report_uri setting below.
```
I wonder if it's going to be confusing to combine this setting with the fact that report_only is always true in development mode. Would it make sense to tie the configurations with the modes here more explicitly, so the user could set the behavior for development or production mode independently? For example, we could make this an array with `report_only[development]` and `report_only[production]` and then the code could test on something like `$this->config->CSP->report_only[APPLICATION_ENV] ?? false` to check the setting.

xmorave2 replied:
I refactored it that way. It is much more transparent what is going on. Thanks for that suggestion.

demiankatz commented:
Thanks for the progress, @xmorave2, I agree that it makes sense to go forward with the custom header code in VuFind for now, and refactor it in the future when your PR gets merged.

I've left a few code-specific comments above; also, some general questions and notes:

1.) I see that a lot of templates have been touched to eliminate inline Javascript. Were you able to test all of these? If not, please let me know and I'll be happy to help set up test scenarios to double-check that nothing got broken.

2.) I think it might make sense to commit this in a few pieces, since a lot of the refactoring to eliminate inline Javascript is not directly related to the CSP logic. Perhaps we can chip away at the size of this PR by merging chunks of refactoring in advance of the actual CSP functionality. Do you have any concerns/objections to that approach?

3.) It looks like this is set up to turn on CSP by default for production. I imagine that may cause some problems for some users, but perhaps that is actually a good thing, since it will highlight security issues. Do you think the best approach is to merge this as-is, and include a note in the changelog indicating that CSP may cause problems, and it can be turned off as needed (but resolving the problems is the better solution)?

4.) Because there's a lot of template/Javascript work here, I'm going to ask for a review from @crhallberg in case he has additional feedback on the approach.

crhallberg commented on themes/bootstrap3/templates/Recommend/SideFacets/cluster-list.phtml:
```diff
@@ -43,7 +43,20 @@
+});
+JS;
+?>
+<?=$this->inlineScript(\Laminas\View\Helper\HeadScript::SCRIPT, $script, 'SET');
```
This is a much cleaner style than the inline `onclick`s, I thank you for that.

I'm concerned that adding event listeners at the end of every facet will duplicate listeners on earlier facets. Can we centralize this in one spot at the end of all the side facets?

xmorave2 replied:
I will continue with that on #1639

crhallberg commented on themes/bootstrap3/templates/devtools/language.phtml:
```diff
@@ -15,8 +15,8 @@
-      <td><?=$missingCount ?><?php if($missingCount > 0): ?> (<a href="#" onclick="return diffManager.showMissing('<?=$langCode ?>')"><?=$this->transEsc('show') ?></a>)<?php endif; ?></td>
-      <td><?=$extraCount ?><?php if($extraCount > 0): ?> (<a href="#" onclick="return diffManager.showExtra('<?=$langCode ?>')"><?=$this->transEsc('show') ?></a>)<?php endif; ?></td>
+      <td><?=$missingCount ?><?php if($missingCount > 0): ?> (<a class="missing_count" data-langcode="<?=$langCode?>" href="#"><?=$this->transEsc('show') ?></a>)<?php endif; ?></td>
+      <td><?=$extraCount ?><?php if($extraCount > 0): ?> (<a class="extra_count" data-langcode="<?=$langCode?>" href="#"><?=$this->transEsc('show') ?></a>)<?php endif; ?></td>
```
These should probably be turned into buttons instead of href-free anchors.

xmorave2 replied:
I will do that on #1639

xmorave2 commented:
@demiankatz @crhallberg Thanks for your reviews!

To your questions @demiankatz:
Ad 1) I cherry-picked the changes in templates/javascript into new PR: #1639, I'll create todo list there. Only change to fie Recommend/VisualFacets.phtml was not moved there, as it depends on cspNonce view helper. (Maybe better to get rid of this view helper, but in this case it would need more refactoring - maybe in own PR?)

Ad 2) Changes into javascript are in own PR, if you think think there are some other candidates for own PR, I have no objections on that.

Ad 3) There is no ability to not use CSP at all, you can only set it into report-only mode - that is generally the right way to implement CSP - set desired policy into report only mode, polish application/policy settings and disable report only mode. The default setting should work probably out of the box for clean VuFindu, but custom templates could broke it easily. The best way would be that for new instalations enable CSP enable by default and for upgrades set up repor-only mode. But it is currently not reachable in VuFind, rigth?

demiankatz commented:
Thanks for the progress, @xmorave2! To respond to your comments:

1) I think it's reasonable to leave the VisualFacets.phtml change here; as long as the majority are migrated to #1639, that will greatly improve readability here.

2) I'll share further thoughts on #1639, as needed.

3) As you say, there's not currently a mechanism to have different defaults for new installations vs. upgrades. I think the easiest approach is to simply choose a default one way or the other, and then include a high priority note in the changelog about it. I can send out a poll to the vufind-tech mailing list when we're a little closer to being ready to merge.

I have an unusually busy week ahead, but I'll try to make some progress on reviewing #1639 as soon as I can.

demiankatz commented:
@xmorave2, looks like your laminas-http PR has just been merged and released as part of v. 2.12.0 of the library. If you have time, you should be able to update this PR to eliminate the need for the local custom HTTP extension. I may be able to help with this later tomorrow if you aren't able to get there first. :-)

xmorave2 commented:
@demiankatz Great news, I will do that soon,

xmorave2 commented:
I merged master into this PR, updated laminas-http and removed custom code. Loooks like it is still working ;)

demiankatz commented on themes/bootstrap3/templates/Recommend/VisualFacets.phtml:
```diff
@@ -31,7 +31,7 @@
 
 <?php if (isset($visualFacetSet)): ?>
-  <script type="text/javascript">
+  <script type="text/javascript" nonce="<?=$this->cspnonce()?>">
```
Would it be better to use the inlineScript helper here?

xmorave2 replied:
definitaly, I've done it

xmorave2 replied:
as it was last occurence of using CspNonce view helper, I even removed that helper

demiankatz commented:
Thanks, @xmorave2; I believe that all of the template/Javascript work is now merged to master at this point! I'll focus on testing the remaining logic later today.

demiankatz commented on config/vufind/contentsecuritypolicy.ini:
```diff
@@ -0,0 +1,32 @@
+default-src[] = "'none'"
+script-src[] = "'strict-dynamic'"
+script-src[] = "'unsafe-inline'"
+script-src[] = "http:"
```
Should we suggest some best practices either here or in a new wiki page? For example, if I'm understanding correctly, many users might want to disable 'http:' if they're running secure sites, etc. I'm happy to establish a wiki page for this if that would be helpful (and then we could link to it from comments here).

xmorave2 replied:
I do think the wiki page would be great for this, could you create one? 

I tried to make the setting as compatible with all kinds of installations and user agents as possible, so there is still room for improve the security

demiankatz replied:
I have created a placeholder page [here](https://vufind.org/wiki/administration:security:content_security_policy). Please feel free to fill in more details. I will also expand it further (with links to GitHub for the .ini file, etc.) after this PR is merged.

demiankatz commented on config/vufind/contentsecuritypolicy.ini:
```diff
@@ -0,0 +1,32 @@
+base-uri[] = "'self'"
+; Set URI which browser should use to report CSP violation; you should provide URL
+; when you enable report_only
+;report-uri[] = 'https://example.report-uri.com'
```
The [documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) suggests that report-uri has been deprecated in favor of report-to, and that best practice is currently to populate both. Is the framework already accounting for this normalization, or do we need to make adjustments here?

xmorave2 replied:
That is true, I do not know why I mixed it this way, I've fixed it

demiankatz commented on config/vufind/contentsecuritypolicy.ini:
```diff
@@ -0,0 +1,32 @@
+script-src[] = "https:"
+connect-src[] = "'self'"
+style-src[] = "'self'"
+img-src[] = "'self'"
```
Is this very string img-src setting going to block direct cover URLs as being developed in #1566? I'm wondering if we need to loosen this up by default, and then recommend that people make it more strict if necessary.

xmorave2 replied:
It i going to block direct cover url.... I do think we could add an example (as I did for MapSelection now) and a note to ObakyKnih setting in ini files... I'll to that on #1566, when this is merged. Is that OK from your point of view?

demiankatz replied:
Yes, that makes sense to me!

demiankatz replied:
I've added a TODO checkbox on #1566 so we don't forget to do this.

demiankatz commented:
@xmorave2, I have a few questions about default configuration and documentation (see above), but I otherwise think this is ready to merge. I should be able to do that next week after we discuss these points. I also see that you have an unchecked to-do item for "Eliminate inline styles." Is there still action needed there?

Thanks again for all of the work on this; this is a great step forward for VuFind's security!

xmorave2 commented:
I've done a little changes here again:
- fixed report-uri/report-to issue in config
- added example values for MapSelection recommendation module into contentsecuritypolicy.ini
- Change default mode to report_only also for production - I do think, we could have report only default in this release and in some next release we will change report_only to false again, what do you think about it @demiankatz

xmorave2 commented:
@demiankatz: regarding the inline styles: I thought we could eliminate them, but it is not possible for some occurancies (in citation templates for example, as then does not work copying with style from web page to text editor). We could remove some of them (maybe most of them), but I am leaving it for another PR.

for now I added unsafe-inline to style-src as default setting.

I also add check to bootstrapper whether is vufind running as script from CLI and not add a CSP header then, as it does not make sense

demiankatz commented:
Thanks for all the work, @xmorave2 -- I agree with your decisions. Making it report only by default is probably safer for now, and we can revisit the decision after we have gained a bit more experience. In any case, we can certainly recommend that people turn it on, but this way we won't frustrate anyone by unexpectedly breaking their installation.

I'll try to get this tested and merged later today when I have a bit more time. More soon!

demiankatz commented:
Okay, @xmorave2, I think this is ready to go. I'll send an email out to vufind-tech to get more eyes on this; I imagine it's possible there may be further discussion of default settings, but I think this is a good baseline to work from! I've also gone ahead and filled in the wiki page I created, but I would welcome further additions and adjustments if you have time.

I'll look further at #1566 once you've had a chance to merge in master and make related updates. In the meantime, I have a few other PRs to work my way through!

Thanks again for all the work on this!

EreMaijala commented:
@xmorave2 I'm trying to set up a permissive CSP and found some issues with the defaults in contentsecuritypolicy.ini (this is according to what Firefox reports):

1. 'strict-dynamic' overrides 'unsafe-inline'
2. use_nonce = true overrides 'unsafe-inline'
3. Nit: "recommend" should be "recommended"

Is there a difference in interpretation between browsers or am I missing something?

EreMaijala commented:
4. It doesn't seem to be possible to disable CSP. Leaving all directives out results in an empty header.

xmorave2 commented:
@EreMaijala: 'unsafe-inline' is set for better compatibility with older/buggy browsers. Current modern browser just ignore it, so it is no problem to have this value set.

demiankatz commented:
Thanks, @EreMaijala for the feedback and @xmorave2 for the clarification. I have corrected the "recommend" typo in the .ini comments. I see that @EreMaijala has opened #1666 to make CSP configuration more flexible. I suggest that we continue further discussions over there.

EreMaijala commented:
Thanks @xmorave2. I added a comment about 'unsafe-inline' in #1666.
